hoverkite-protocol/src/response.rs

```diff
@@ -0,0 +1,189 @@
+                    side,
+                    response: SideResponse::BatteryReadings {
+                        battery_voltage,
+                        backup_battery_voltage,
+                        motor_current,
+                    },
+                })
+            } else {
+                None
+            }),
+            Some(b'C') | Some(b'c') => Ok(if buffer.len() >= 2 {
+                let side = if buffer.pop_front().unwrap() == b'C' {
+                    side
+                } else {
+                    side.opposite()
+                };
+                let byte = buffer.pop_front().unwrap();
+                let charger_connected = match byte {
+                    b'0' => false,
+                    b'1' => true,
+                    _ => return Err(UnexpectedResponse(byte)),
+                };
+                Some(Response {
+                    side,
+                    response: SideResponse::ChargeState { charger_connected },
+                })
+            } else {
+                None
+            }),
+            Some(r) => {
+                buffer.pop_front();
+                Err(UnexpectedResponse(r))
+            }
+            None => Ok(None),
+        }
+    }
+}
+
+#[cfg(feature = "std")]
+#[cfg(test)]
+mod tests {
+    use super::*;
+    use test_case::test_case;
+
+    #[test]
+    fn parse_invalid() {
+        let mut buffer = VecDeque::new();
+        buffer.extend(b"x");
+        assert_eq!(
+            Response::parse(&mut buffer, Side::Right),
+            Err(UnexpectedResponse(b'x'))
+        );
+        assert_eq!(buffer.len(), 0);
+    }
+
+    #[test]
+    fn parse_empty() {
+        let mut buffer = VecDeque::new();
+        assert_eq!(Response::parse(&mut buffer, Side::Right), Ok(None));
+        assert_eq!(buffer.len(), 0);
+    }
+
+    // TODO: test_case doesn't allow test cases which differ only in case.
+    #[test_case(b"I")]
+    #[test_case(b"B12345")]
+    #[test_case(b"C")]
+    #[test_case(b"\"blah")]
+    fn parse_partial(partial_response: &[u8]) {
+        for length in 1..=partial_response.len() {
+            let mut buffer = VecDeque::new();
+            buffer.extend(&partial_response[..length]);
+            assert_eq!(Response::parse(&mut buffer, Side::Right), Ok(None));
+            // No bytes should be consumed from the buffer.
+            assert_eq!(buffer.len(), length);
+        }
+    }
+
+    #[test]
+    fn parse_invalid_charge_state() {
+        let mut buffer = VecDeque::new();
+        buffer.extend(b"Cx");
+        assert_eq!(
+            Response::parse(&mut buffer, Side::Right),
+            Err(UnexpectedResponse(b'x'))
+        );
+        assert_eq!(buffer.len(), 0);
+    }
+
+    #[test_case(b"I12345678", SideResponse::Position(4050765991979987505))]
+    #[test_case(b"B123456", SideResponse::BatteryReadings {
+        battery_voltage: 12849,
+        backup_battery_voltage: 13363,
+        motor_current: 13877,
```

Comment:
Maybe use hex here and bye array syntax above so it's more obvious that the numbers match up??

Reply:
Done.